sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -0,0 +1,274 @@
+      .filter(e => e.references.subsetOf(curPlan.outputSet ++ otherPlan.outputSet))
+
+    if (joinCond.nonEmpty) {
+      val curPlanStats = curPlan.stats(conf)
+      val otherPlanStats = otherPlan.stats(conf)
+      if (curPlanStats.rowCount.nonEmpty && otherPlanStats.rowCount.nonEmpty) {
+        // Now both curPlan and otherPlan become intermediate joins, so the cost of the
+        // new join should also include their costs.
+        val cost = curJoinPlan.cost + otherJoinPlan.cost +
+          Cost(curPlanStats.rowCount.get, curPlanStats.sizeInBytes) +
+          Cost(otherPlanStats.rowCount.get, otherPlanStats.sizeInBytes)
+
+        // Put the deeper side on the left, tend to build a left-deep tree.
+        val (left, right) = if (curJoinPlan.itemIds.size >= otherJoinPlan.itemIds.size) {
+          (curPlan, otherPlan)
+        } else {
+          (otherPlan, curPlan)
+        }
+        val newJoin = Join(left, right, Inner, joinCond.reduceOption(And))
+        val remainingConds = conditions -- collectJoinConds(newJoin)
+        val neededAttr = AttributeSet(remainingConds.flatMap(_.references)) ++ topOutput
+        val newPlan =
+          if ((newJoin.outputSet -- newJoin.outputSet.filter(neededAttr.contains)).nonEmpty) {
+            Project(newJoin.output.filter(neededAttr.contains), newJoin)
+          } else {
+            newJoin
+          }
+        val itemIds = curJoinPlan.itemIds.union(otherJoinPlan.itemIds)
+        return Some(JoinPlan(itemIds, newPlan, cost))
+      }
+    }
+    None
+  }
+
+  private def collectJoinConds(plan: LogicalPlan): Set[Expression] = plan match {
+    case j @ Join(left, right, _: InnerLike, cond) =>
+      val leftConditions = collectJoinConds(left)
+      val rightConditions = collectJoinConds(right)
+      cond.map(splitConjunctivePredicates).getOrElse(Nil).toSet ++ leftConditions ++ rightConditions
+    case Project(_, j @ Join(left, right, _: InnerLike, cond)) =>
+      val leftConditions = collectJoinConds(left)
+      val rightConditions = collectJoinConds(right)
+      cond.map(splitConjunctivePredicates).getOrElse(Nil).toSet ++ leftConditions ++ rightConditions
+    case _ =>
+      Set()
+  }
+
+  /**
+   * Partial join order in a specific level.
+   *
+   * @param itemIds Set of item ids participating in this partial plan.
+   * @param plan The plan tree with the lowest cost for these items found so far.
+   * @param cost The cost of this plan is the sum of costs of all intermediate joins.
+   */
+  case class JoinPlan(itemIds: Set[Int], plan: LogicalPlan, cost: Cost)
+}
+
+/** This class defines the cost model. */
+case class Cost(rows: BigInt, sizeInBytes: BigInt) {
+  /**
+   * An empirical value for the weights of cardinality (number of rows) in the cost formula:
+   * cost = rows * weight + size * (1 - weight), usually cardinality is more important than size.
+   */
+  val weight = 0.7
+
+  def +(other: Cost): Cost = Cost(rows + other.rows, sizeInBytes + other.sizeInBytes)
```

Comment:
Sorry can you give me a clue about the style? Is it about the method name? I see `BitSet.scala` has something similar like `def &(other: BitSet): BitSet`.